### ISSUE TITLE
Compile for higher GPU architectures

### ISSUE THREAD

**Issue by f0k** [cudamat/cudamat#18]
Changing the Makefile to match the compute capabilities of the device at hands can provide performance benefits. For example, on a GTX 580 (compute capability 2.0), adding `-arch=sm_20` to the `nvcc` calls consistently improves performance of element-wise addition and multiplication by 8% for small matrices (both with and without the dynamic blocksize feature). The current Makefile always compiles for compute capabilities 1.0.

We should consider adding a configure script to detect (or manually select) all compute capabilities to compile for, or even switch to a build system for the package.

**Comment by Rietaros** [comment]
I'm copiling this code : 
```

libcudamat.so: cudamat.cu cudamat_kernels.cu cudamat.cuh cudamat_kernels.cuh
	nvcc -O3 \
		-v \
		-gencode=arch=compute_10,code=sm_10 \
	  -gencode=arch=compute_20,code=sm_20 \
		-gencode=arch=compute_30,code=sm_30 \
		--compiler-options '-fPIC' -o libcudamat.so \
		--shared cudamat.cu cudamat_kernels.cu -lcublas -L$(CUDA_LIB)

libcudamat_conv.so: cudamat_conv.cu cudamat_conv_kernels.cu cudamat_conv.cuh cudamat.cuh cudamat_conv_kernels.cuh
	nvcc -O3 \
		-v \
		-gencode=arch=compute_10,code=sm_10 \
	  -gencode=arch=compute_20,code=sm_20 \
		-gencode=arch=compute_30,code=sm_30 \
		--compiler-options '-fPIC' -o libcudamat_conv.so \archite
		--shared cudamat_conv.cu cudamat_conv_kernels.cu -lcublas -L$(CUDA_LIB)
```

give me the output of wrong architecture,
I'm using NVIDIA Geforce 940M, cuda 7.5, nvidia-352.39.
How I get the right architecture?

I've read someone use : 

```
libcudamat.so: cudamat.cu cudamat_kernels.cu cudamat.cuh cudamat_kernels.cuh
	nvcc -O3 \
		-v \
	  -gencode=arch=compute_20,code=sm_20 \
		-gencode=arch=compute_30,code=sm_30 \
		-gencode=arch=compute_35,code=sm_35 \
		-gencode=arch=compute_50,code=sm_50 \
		--compiler-options '-fPIC' -o libcudamat.so \
		--shared cudamat.cu cudamat_kernels.cu -lcublas -L$(CUDA_LIB)

libcudamat_conv.so: cudamat_conv.cu cudamat_conv_kernels.cu cudamat_conv.cuh cudamat.cuh cudamat_conv_kernels.cuh
	nvcc -O3 \
		-v \
	  -gencode=arch=compute_20,code=sm_20 \
		-gencode=arch=compute_30,code=sm_30 \
		--compiler-options '-fPIC' -o libcudamat_conv.so \
		--shared cudamat_conv.cu cudamat_conv_kernels.cu -lcublas -L$(CUDA_LIB)
```

but it's make gpu_lock.obtain_lock_id() always give -1

Any suggestion?
thank you

**Comment by Rietaros** [comment]
also I find this error on my 2nd compile

```
libcudamat.so libcudamat_conv.so
make: libcudamat.so: Command not found
Makefile:3: recipe for target 'all' failed
make: *** [all] Error 127
```

**Comment by f0k** [comment]
> I'm using NVIDIA Geforce 940M Maxwell architevture, cuda 7.5, nvidia-352.39. How I get the right architecture?

As per [a table maintained on Wikipedia](https://en.wikipedia.org/wiki/CUDA#Supported_GPUs), a 940M has compute capability 5.0. So as per the [installation instructions](https://github.com/cudamat/cudamat/blob/master/INSTALL.md), you can install it with `NVCCFLAGS=-arch=sm_50 python setup.py install --user`. Note that if you've got a Makefile, then you're not using cudamat from [this very repository](/cudamat/cudamat). It doesn't have a Makefile any longer.

**Comment by f0k** [comment]
> I guess the only way to detect the compute capability at configure time is by compiling and running a small program calling cudaGetDeviceProperties() for every device.

No. One can enumerate the devices and query their compute capabilities using `ctypes`, without compiling anything (demo: https://gist.github.com/f0k/63a664160d016a491b2cbea15913d549). Doing this from within `setup.py` would allow us to set the nvcc flags automatically.
